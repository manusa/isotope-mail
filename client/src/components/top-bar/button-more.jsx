import React, {useState, useEffect} from 'react';
import {translate} from 'react-i18next';
import TopBarButton from './top-bar-button';
import mainCss from '../../styles/main.scss';

export const ButtonMore = ({t, children}) => {
  const [menuOpen, setMenuOpen] = useState(false);
  useEffect(() => {
    function closeMenu() {
      setMenuOpen(false);
    }
    window.addEventListener('click', closeMenu);
    return function unmount() {
      window.removeEventListener('click', closeMenu);
    };
  }, []);
  return (
    <span
      className={`${mainCss['mdc-menu-surface--anchor']}`}
      isotip={t('topBar.more')} isotip-position='bottom-end' isotip-size='small'
      isotip-hidden={menuOpen.toString()}>
      <TopBarButton
        onClick={event => {
          setMenuOpen(!menuOpen);
          event.stopPropagation();
        }}>more_vert</TopBarButton>
      {React.cloneElement(children, {visible: menuOpen})}
    </span>
  );
};

export default translate()(ButtonMore);
