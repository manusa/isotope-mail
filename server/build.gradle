buildscript {
	ext {
		springBootVersion = '2.0.4.RELEASE'
	}
	repositories {
		mavenCentral()
	}
	dependencies {
		classpath("org.springframework.boot:spring-boot-gradle-plugin:${springBootVersion}")
	}
}

apply plugin: 'java'
apply plugin: 'eclipse'
apply plugin: 'org.springframework.boot'
apply plugin: 'io.spring.dependency-management'

group = 'com.marcnuri.isotope'
version = '0.0.1-SNAPSHOT'
sourceCompatibility = 1.8

repositories {
	mavenCentral()
}


dependencies {
	compile('org.springframework.boot:spring-boot-starter')
	compile('org.springframework.boot:spring-boot-starter-web')
	compile('org.springframework.boot:spring-boot-starter-hateoas')
	compile('com.sun.mail:javax.mail:1.6.1')
	testCompile('org.springframework.boot:spring-boot-starter-test')
	testCompile('com.marcnuri:spring-common-test:0.0.2')
}
