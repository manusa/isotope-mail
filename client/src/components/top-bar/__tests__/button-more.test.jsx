import React from 'react';
import {act} from 'react-dom/test-utils';
import {shallow, mount} from 'enzyme/build/index';
import {ButtonMore} from '../button-more';

describe('ButtonMore component test suite', () => {
  let t;
  beforeAll(() => {
    t = jest.fn(arg => arg);
  });

  describe('Snapshot render', () => {
    test('should render button-more', () => {
      // Given
      const props = {t};
      // When
      const buttonMore = shallow(<ButtonMore {...props}><div>Child</div></ButtonMore>);
      // Then
      expect(buttonMore).toMatchSnapshot();
    });
  });
  describe('Events tests', () => {
    test('click, should set children visible', () => {
      // Given
      const props = {t};
      const event = {stopPropagation: jest.fn()};
      const child = <div>Child</div>;
      const buttonMore = shallow(<ButtonMore {...props}>{child}</ButtonMore>);
      // When
      buttonMore.find('TopBarButton').props().onClick(event);
      // Then
      expect(event.stopPropagation).toHaveBeenCalledTimes(1);
      expect(buttonMore.props()['isotip-hidden']).toBe('true');
    });
    test('unmount, window should not contain event listeners', () => {
      // Given
      const originalAddEventListener = window.addEventListener;
      const originalRemoveEventListener = window.removeEventListener;
      window.addEventListener = jest.fn(() => originalAddEventListener.apply(null, arguments));
      window.removeEventListener = jest.fn(() => originalRemoveEventListener.apply(null, arguments));
      const props = {t};
      const buttonMore = mount(<ButtonMore {...props}><div>Child</div></ButtonMore>);
      // When
      buttonMore.unmount();
      // Then
      expect(window.addEventListener).toHaveBeenCalledWith('click', expect.any(Function));
      expect(window.removeEventListener).toHaveBeenCalledWith('click', expect.any(Function));
    });
    test('window click, menu close event should be fired', () => {
      // Given
      const props = {t};
      const buttonMore = mount(<ButtonMore {...props}><div>Child</div></ButtonMore>);
      act(() => {
        buttonMore.find('TopBarButton').props().onClick(new Event('click')); // Set menu Open true
      });
      // When
      act(() => {
        window.dispatchEvent(new Event('click')); // Close menu
      });
      // Then
      expect(buttonMore.find('span').props()['isotip-hidden']).toBe('false');
    });
  });
});
