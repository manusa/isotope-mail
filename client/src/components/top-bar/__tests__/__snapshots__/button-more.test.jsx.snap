// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonMore component test suite Snapshot render should render button-more 1`] = `
<span
  className="mdc-menu-surface--anchor"
  isotip="topBar.more"
  isotip-hidden="false"
  isotip-position="bottom-end"
  isotip-size="small"
>
  <TopBarButton
    className=""
    onClick={[Function]}
  >
    more_vert
  </TopBarButton>
  <div
    visible={false}
  >
    Child
  </div>
</span>
`;
